### PR TITLE
python-ndn: init at 0.3-3

### PR DESCRIPTION
**Pull request by Janik-Haag**
###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/networking/pysim/default.nix [hunk 1]
```diff
@@ -0,0 +1,43 @@
+{ lib
+, python3
+, fetchFromGitHub
+, makeDesktopItem
+, copyDesktopItems
+}:
+
+python3.pkgs.buildPythonApplication rec {
+  pname = " pysim";
```
Comment (sikmir) [inline pkgs/tools/networking/pysim/default.nix:9]:
```suggestion
  pname = "pysim";
```

File: pkgs/tools/networking/pysim/default.nix [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  version = "unstable-2023-05-27";
+
+  src = fetchFromGitHub {
+    owner = "osmocom";
+    repo = pname;
+    rev = "c296cb593e9bc27ab6d482b88b007db51d5eaa3f";
+    hash = "";
```
Comment (sikmir) [inline pkgs/tools/networking/pysim/default.nix:16]: Fill hash.

File: pkgs/tools/networking/pysim/default.nix [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+, copyDesktopItems
+}:
+
+python3.pkgs.buildPythonApplication rec {
+  pname = " pysim";
+  version = "unstable-2023-05-27";
+
+  src = fetchFromGitHub {
+    owner = "osmocom";
+    repo = pname;
```
Comment (sikmir) [inline pkgs/tools/networking/pysim/default.nix:14]:
```suggestion
    repo = "pysim";
```

File: pkgs/tools/networking/pysim/default.nix [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    bidict
+    gsm0338
+    pyyaml
+    termcolor
+    colorlog
+    packaging
+    pycryptodome
+  ];
+
+  #doCheck = false;
```
Comment (sikmir) [inline pkgs/tools/networking/pysim/default.nix:35]:
```suggestion
```

File: pkgs/tools/networking/pysim/default.nix [hunk 5]
```diff
@@ -0,0 +1,43 @@
+{ lib
+, python3
+, fetchFromGitHub
+, makeDesktopItem
+, copyDesktopItems
```
Comment (sikmir) [inline pkgs/tools/networking/pysim/default.nix:5]:
```suggestion
```
You don't use it.

**Comment by Janik-Haag** [comment]
Wait I messed up.. give me a few minuts.

**Comment by Janik-Haag** [comment]
I accidentally committed the wrong file, pysim is still a work in progress on my local machine :see_no_evil: 
It's fixed now, sorry for wasting your time.

File: pkgs/development/python-modules/python-ndn/default.nix [hunk 6]
```diff
@@ -0,0 +1,36 @@
+{ pkgs
```
Comment (sikmir) [inline pkgs/development/python-modules/python-ndn/default.nix:1]:
```suggestion
{ lib
```

File: pkgs/development/python-modules/python-ndn/default.nix [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  propagatedBuildInputs = with python3Packages; [
+    pycryptodomex
+    lark
+    pygtrie
+    aenum
+  ];
+
+  pythonImportChecks = [ "ndn" ];
+
+  meta = with pkgs.lib; {
```
Comment (sikmir) [inline pkgs/development/python-modules/python-ndn/default.nix:30]:
```suggestion
  meta = with lib; {
```

**Comment by sikmir** [comment]
Result of `nixpkgs-review pr 234582` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>python310Packages.python-ndn</li>
    <li>python310Packages.python-ndn.dist</li>
  </ul>
</details>

**Review by sikmir — APPROVED** [review]

File: pkgs/development/python-modules/python-ndn/default.nix [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+, buildPythonPackage
+, fetchFromGitHub
+, python3Packages
+, pythonOlder
+}:
+
+buildPythonPackage rec {
+  pname = "python-ndn";
+  version = "0.3-1";
+
```
Comment (natsukium) [inline pkgs/development/python-modules/python-ndn/default.nix:11]:
```suggestion
format = "pyproject";

```

File: pkgs/development/python-modules/python-ndn/default.nix [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+  src = fetchFromGitHub {
+    owner = "named-data";
+    repo = pname;
+    rev = "v${version}";
+    hash = "sha256-+kWhUNu65yQMHP4m/86j0XqrZGrvtkYOLeLKXjzRhpY=";
+  };
+
+  disabled = pythonOlder "3.8";
+
+  propagatedBuildInputs = with python3Packages; [
```
Comment (natsukium) [inline pkgs/development/python-modules/python-ndn/default.nix:21]:
```suggestion
  propagatedBuildInputs = [
```

File: pkgs/development/python-modules/python-ndn/default.nix [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    pycryptodomex
+    lark
+    pygtrie
+    aenum
+  ];
+
```
Comment (natsukium) [inline pkgs/development/python-modules/python-ndn/default.nix:27]:
```suggestion

nativeCheckInputs = [
  pytestCheckHook
];

```

File: pkgs/development/python-modules/python-ndn/default.nix [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+  pythonImportChecks = [ "ndn" ];
+
+  meta = with lib; {
```
Comment (natsukium) [inline pkgs/development/python-modules/python-ndn/default.nix:30]: Would you please add `meta.changelog`?

File: pkgs/development/python-modules/python-ndn/default.nix [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+  version = "0.3-1";
+
+  src = fetchFromGitHub {
+    owner = "named-data";
+    repo = pname;
+    rev = "v${version}";
+    hash = "sha256-+kWhUNu65yQMHP4m/86j0XqrZGrvtkYOLeLKXjzRhpY=";
+  };
+
+  disabled = pythonOlder "3.8";
```
Comment (natsukium) [inline pkgs/development/python-modules/python-ndn/default.nix:19]: It seems that the upstream requires 3.9 or higher.
https://github.com/named-data/python-ndn/blob/1725b13d4c34afe1dca6615b6338bb3edb846c88/setup.cfg#L37

Comment (Janik-Haag) [reply pkgs/development/python-modules/python-ndn/default.nix]: argh, typo

File: pkgs/development/python-modules/python-ndn/default.nix [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+  };
+
+  disabled = pythonOlder "3.9";
+
+  propagatedBuildInputs = [
+    pycryptodomex
+    lark
+    pygtrie
+    aenum
+    setuptools
```
Comment (natsukium) [inline pkgs/development/python-modules/python-ndn/default.nix:34]: setuptools should go to nativeBuildInputs

File: pkgs/development/python-modules/python-ndn/default.nix [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+, pycryptodomex
+, lark
+, pygtrie
+, aenum
+, pytestCheckHook
+}:
+
+buildPythonPackage rec {
+  pname = "python-ndn";
+  version = "0.3-1";
```
Comment (natsukium) [inline pkgs/development/python-modules/python-ndn/default.nix:16]: There are no git tags, but according to PyPI, it looks like the latest release, 0.3-3, is [this commit](https://github.com/named-data/python-ndn/commit/8a852659139e974d7a624b40a9145b4844c9af07).

File: pkgs/development/python-modules/python-ndn/default.nix [hunk 15]
```diff
@@ -0,0 +1,50 @@
+{ lib
+, buildPythonPackage
+, fetchFromGitHub
+, python3Packages
```
Comment (natsukium) [inline pkgs/development/python-modules/python-ndn/default.nix:4]:
```suggestion
```

File: pkgs/development/python-modules/python-ndn/default.nix [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+  nativeCheckInputs = [
+    pytestCheckHook
+  ];
+
+  pythonImportChecks = [ "ndn" ];
+
+  meta = with lib; {
+    description = "An NDN client library with AsyncIO support in Python 3";
+    homepage = "https://github.com/named-data/python-ndn";
+    changelog = "https://github.com/named-data/python-ndn/blob/master/CHANGELOG.rst#03-1-2022-03-20";
```
Comment (natsukium) [inline pkgs/development/python-modules/python-ndn/default.nix:46]:
```suggestion
    changelog = "https://github.com/named-data/python-ndn/blob/${rev}/CHANGELOG.rst;
```
would it be better?

File: pkgs/development/python-modules/python-ndn/default.nix [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+, buildPythonPackage
+, fetchFromGitHub
+, python3Packages
+, pythonOlder
+, setuptools
+, pycryptodomex
+, lark
+, pygtrie
+, aenum
+, pytestCheckHook
```
Comment (fabaff) [inline pkgs/development/python-modules/python-ndn/default.nix:11]: `lib` first and the rest of the inputs alphabetically ordered will make the diffs simpler to read in the future.

```suggestion
, aenum
, buildPythonPackage
, fetchFromGitHub
, lark
, pycryptodomex
, pygtrie
, pytestCheckHook
, python3Packages
, pythonOlder
, setuptools
```

File: pkgs/development/python-modules/python-ndn/default.nix (hunk 13 above)
Comment (fabaff) [reply pkgs/development/python-modules/python-ndn/default.nix:34]: `nativeBuildInputs` will change with the next update  as upstream uses [`poetry-core`](https://github.com/named-data/python-ndn/blob/master/pyproject.toml#L2) in `master`.

File: pkgs/development/python-modules/python-ndn/default.nix [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+  ];
+
+  nativeCheckInputs = [
+    pytestCheckHook
+  ];
+
+  pythonImportChecks = [ "ndn" ];
+
+  meta = with lib; {
+    description = "An NDN client library with AsyncIO support in Python 3";
```
Comment (fabaff) [inline pkgs/development/python-modules/python-ndn/default.nix:44]: Opinionated change

```suggestion
    description = "An NDN client library";
```

File: pkgs/development/python-modules/python-ndn/default.nix (hunk 16 above)
Comment (fabaff) [inline pkgs/development/python-modules/python-ndn/default.nix:46]: My guess is that `nix-update` wouldn't be able to add the changelog to the commit message.

```suggestion
    changelog = "https://github.com/named-data/python-ndn/blob/v${version}/CHANGELOG.rst";
```

**Review by fabaff — COMMENTED** [review]
The latest release seems to be `0.3-3`

https://github.com/named-data/python-ndn/issues/65

File: pkgs/development/python-modules/python-ndn/default.nix (hunk 16 above)
Comment (Janik-Haag) [reply pkgs/development/python-modules/python-ndn/default.nix:46]: Yeah you are right, will change it.

**Comment by Janik-Haag** [comment]
Regarding version and setup tools: I could change the source to pypi and then we would have the newer version and poetry or would you prefer another way @natsukium  @fabaff ?

**Comment by natsukium** [comment]
In this case, it seems that using `fetchPypi` is easier to update because it does not need to be updated rev manually.

File: pkgs/development/python-modules/python-ndn/default.nix [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+  nativeCheckInputs = [
+    pytestCheckHook
+  ];
+
+  pythonImportChecks = [ "ndn" ];
+
+  meta = with lib; {
+    description = "An NDN client library with AsyncIO support";
+    homepage = "https://github.com/named-data/python-ndn";
+    changelog = "https://github.com/named-data/python-ndn/blob/${rev}/CHANGELOG.rst;
```
Comment (natsukium) [inline pkgs/development/python-modules/python-ndn/default.nix:45]:
```suggestion
    changelog = "https://github.com/named-data/python-ndn/blob/${src.rev}/CHANGELOG.rst;
```
Sorry, my mistake.

Comment (Janik-Haag) [reply pkgs/development/python-modules/python-ndn/default.nix]: fixed

**Comment by Janik-Haag** [comment]
I left it with fetchFromGithub because the tags got updated.

**Comment by natsukium** [comment]
Please add `setuptools` to `nativeBuildInputs`, otherwise LGTM.

**Comment by Janik-Haag** [comment]
Should be fine now.

File: pkgs/development/python-modules/python-ndn/default.nix [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+    pycryptodomex
+    lark
+    pygtrie
+    aenum
+    aiohttp
+  ];
+
+  nativeCheckInputs = [
+    pytestCheckHook
+    setuptools
```
Comment (natsukium) [inline pkgs/development/python-modules/python-ndn/default.nix:39]:
```suggestion
```

**Comment by Janik-Haag** [comment]
Sorry, my bad :see_no_evil:

**Review by natsukium — APPROVED** [review]

**Comment by sikmir** [comment]
Result of `nixpkgs-review pr 234582` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.python-ndn</li>
    <li>python310Packages.python-ndn.dist</li>
    <li>python311Packages.python-ndn</li>
    <li>python311Packages.python-ndn.dist</li>
  </ul>
</details>

**Comment by sikmir** [comment]
Result of `nixpkgs-review pr 234582` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.python-ndn</li>
    <li>python310Packages.python-ndn.dist</li>
    <li>python311Packages.python-ndn</li>
    <li>python311Packages.python-ndn.dist</li>
  </ul>
</details>

**Review by sikmir — APPROVED** [review]

**Pull request merged**